ユーザー管理機能

ken-forest opened the pull request:
# What
ユーザー登録できる機能を実装

# Why
ログイン・ログアウトでフリマアプリを利用する
ユーザー情報の整理を行うため。

＜Gyazoで撮影したもの＞
・必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/103148281db1affaf9b05c20ad760ace

・入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/3555716ab46ae128a3584920c7266794

・必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/7758dba57d81b1fcc11ac1b8b11dd865

・入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/eaa9439471d6b711cad97058b6badb7d

・ログアウトに成功する動画
https://gyazo.com/e1faada0407712e7d25174489a8f4556

・テストに成功する画像
https://gyazo.com/3e6041892eed9eba22501ed04de1d9ce

ken-forest committed:
deviseGemの導入

ken-forest committed:
userモデルの作成

ken-forest committed:
userモデルの立ち上げと、モデルにバリデーションの設定

ken-forest committed:
rails g devise:viewsで作成した registrations/new.html.erb と sessions/new.html.erb を、配布済みのビューファイルに置き換えた

ken-forest committed:
新規登録画面へ飛べるようにした

①app/views/shared/_header.html.erb
このファイルの新規登録ボタンのパスを
new_user_registration_pathに変更

②app/views/devise/registrations/new.html.erb
このファイルのform_withのパスを
user_registration_pathに変更し、
モデルの欄に@userを追加

ken-forest committed:
app/views/devise/registrations/new.html.erbのhogeを全て、該当のカラム名に変更

ken-forest committed:
app/controllers/application_controller.rbにストロングパラメータを定義

ken-forest committed:
ログインの有無でトップページのヘッダー右のボタンに変化が出るように変更

ken-forest committed:
ログインとログアウト機能の作成、その他

・ログインボタンにnew_user_session_path
・ログアウトボタンに
destroy_user_session_path
・views/devise/sessions/new.html.erb
のform_withに
model: @user, url: user_session_path
を追加
・views/shared/_header.html.erbの
20と21行目に「%>」が抜けていたので
追加

ken-forest committed:
app/views/devise/sessions/new.html.erbの全てのhogeを該当するカラム名に修正

ken-forest committed:
新規登録時、不備があるとエラーメッセージが出るように、app/views/devise/registrations/new.html.erbの14行目を変更

ken-forest committed:
テストコード、RSpecGemの導入

ken-forest committed:
pry-railsGemの導入

ken-forest committed:
FactoryBotGemの導入

ken-forest committed:
テストコード試作

ken-forest committed:
userモデルのpasswordにバリデーション追加

正規表現を使って
パスワードに、英数字６文字以上という
制限をつけた。

ken-forest committed:
userモデルの４箇所に新たなバリデーション追加

last_nameとfirst_nameに
全角ひらがな・漢字のバリデーション
last_name_kanaとfirst_name_kanaに
全角カタカナのバリデーション
を追加

ken-forest committed:
spec/models/user_spec.rbにuserモデルの単体テストコードを記述

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+         :recoverable, :rememberable, :validatable
+         
+  validates :nickname, presence: true
+  validates :last_name, presence: true, format:{with: /\A[一-龥ぁ-ん]/ }
```
```rb
  validates :last_name, presence: true, format:{with: /\A[一-龥ぁ-ん]/ }
  validates :first_name, presence: true, format:{with: /\A[一-龥ぁ-ん]/ }
```
現状だと、最初の文字にしかバリデーションが適用されない状態となっています。
全ての文字にバリデーションが適用されるよう`z/`の記述を追加しましょう。

TakashiKozuma replied:
現状だと、カタカナや、佐々木の「々」 や市ヶ谷の「ケ」などが登録できないので、登録できるように修正しましょう。

TakashiKozuma commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
```rb
<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
 ```
3,5行目のコメントアウトは実装案内なので削除しておきましょう。
実装案内は完成したアプリを人に見せる際に残しておくの望ましくないためです。
他にも「〜しましょう」のような実装案内があれば、実装時に削除するようにしていきましょう。

TakashiKozuma commented on app/views/shared/_header.html.erb:
```diff
@@ -16,10 +16,13 @@
+      <li><%= link_to 'ログイン', new_user_session_path, class: "login" %></li>
+      <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
+      <% end %>
       <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
```rb
<%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
18,26行目も実装案内なので、削除しておきましょう。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

ken-forest committed:
app/views/devise/registrations/new.html.erbの３行目・５行目のコメントアウトの実装案内を削除

ken-forest committed:
userモデルのlast_nameとfirst_nameのバリデーションをを修正

両方のバリデーションの正規表現の部分後半に、「\z」の記述を追加

ken-forest committed:
userモデルのlast_nameとfirst_nameのバリデーションを変更

具体的には、入力する時に、
全角ひらがな・全角カタカナ・漢字を許可するように変更。
正規表現を以下のように上書きした。

/\A(?:\p{Hiragana}|\p{Katakana}|[ー－]|[一-龠々])+\z/

ken-forest committed:
app/views/shared/_header.html.erbの18行目と26行目にあった実装案内を削除

ken-forest commented:
ご指摘いただいた、４つの項目を修正させていただきました。
引き続き、実装内容に不備が見つかれば、コメントをよろしくお願いいたします！

shodai-ito commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :first_name_kana, presence: true, format:{with: /\A[ァ-ヶー－]+\z/ }
+  validates :birth_date, presence: true
+
+  validates :password, format:{with: /\A[a-z\d]{6,}\z/ }
```
`validates :password, format:{with: /\A[a-z\d]{6,}\z/ }`
こちらの正規表現ですが、「半角英数字のみ」という意味なります。
逆に言えば、「半角英数字であれば英字のみでも数字のみでも可」となってしまいます。
今回の実装は「半角英数字混合が必須」ですので、正規表現を書き換えましょう！

shodai-ito commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%= render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
実装内容を指示するコメントアウトは削除しましょう！

shodai-ito commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
+  factory :user do
+    nickname              {'yamada'}
+    email                 {'yamada@gmail.com'}
```
`email                 {'yamada@gmail.com'}`
emailのような一意性の項目については、Fakerを用いてダミーデータを作成されることをおすすめします！

shodai-ito commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,88 @@
+
+    it 'nickname、email、passwordとpassword_confirmation、last_name、first_name、last_name_kana、first_name_kana、birth_dateが存在すれば登録できる' do
+      expect(@user).to be_valid
+    end
```
テストコードはcontextを用いて「正常系」と「異常系」に分けましょう！

shodai-ito commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,88 @@
+      @user.password_confirmation = '00000'
+      @user.valid?
+      expect(@user.errors.full_messages).to include('Password is too short (minimum is 6 characters)')
+    end
```
以下のテストが必要です。
追加しましょう！
```
・メールアドレスに@を含まない場合は登録できない
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない
・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
```

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
追加でコメントしましたので、ご確認をお願いいたします！

その後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

ken-forest committed:
FakerGemの導入

ken-forest committed:
spec/factories/users.rbのemailの定義部分をFakerを使った記述に変更

ken-forest committed:
app/views/devise/registrations/new.html.erbの13と15行目の実装案内を削除

ken-forest committed:
spec/models/user_spec.rbにcontextの記述を使い、正常系・異常系を分けた

ken-forest committed:
バリデーション修正、spec/factories/users.rbのpassword定義の中身を修正、その他

userモデルのpasswordのバリデーションを修正。以下の正規表現を上書き。
「/\A(?=.*?[a-z])(?=.*?[\d])\w{6,}\z/」
これで、半角6文字以上でアルファベット小文字・数字それぞれ１文字以上含む必要、、という条件になる。
その条件に伴うように、
spec/factories/users.rbのpassword定義の中身を変更。
最後にspec/models/user_spec.rbに
・メールアドレスに@を含まない場合は登録できない
・英字のみのパスワードでは登録できない
以上２つのテストコードを追加

ken-forest committed:
userモデルのバリデーションを変更、その他

app/models/user.rbのlast_nameとfirst_nameのバリデーションを
「/[^\x01-\x7E\uFF61-\uFF9F]/」に
変更。
spec/models/user_spec.rbに
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない
以上のテストコードを追加

ken-forest committed:
spec/models/user_spec.rbにテストコードを追加

・last_name_kanaがひらがなでは登録できない
・last_name_kanaが漢字では登録できない
・last_name_kanaがアルファベットでは登録できない
・last_name_kanaが数字では登録できない
・last_name_kanaが記号では登録できない
以上５つのテストコードを追加

ken-forest committed:
spec/models/user_spec.rbにテストコードを追加

・first_name_kanaがひらがなでは登録できない
・first_name_kanaが漢字では登録できない
・first_name_kanaがアルファベットでは登録できない
・first_name_kanaが数字では登録できない
・first_name_kanaが記号では登録できない
以上５つのテストコードを追加

ken-forest commented:
ご指摘いただいた箇所を修正させていただきました。
引き続き、レビューをよろしくお願いします。

＜Gyazoで撮影した新たなテストコード＞
https://gyazo.com/55ddb2a0aea1c2a2a91c7bf8f15ea6f5

mihot117 commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+         :recoverable, :rememberable, :validatable
+         
+  validates :nickname, presence: true
+  validates :last_name, presence: true, format:{with: /[^\x01-\x7E\uFF61-\uFF9F]/ }
+  validates :first_name, presence: true, format:{with: /[^\x01-\x7E\uFF61-\uFF9F]/ }
```
```ruby
validates :last_name, presence: true, format:{with: /[^\x01-\x7E\uFF61-\uFF9F]/ }
  validates :first_name, presence: true, format:{with: /[^\x01-\x7E\uFF61-\uFF9F]/ }
```

上記の記述ですと、全角数字も入力できてしまう可能性がございます。
元々記載いただいていた`format:{with: /\A[一-龥ぁ-ん]/ }`の記述を少し修正いただくだけで実装が可能となりますので、以下のように修正をお願いいたします。

```ruby
validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/ }
```

`々`の記述があることによって、佐々木さんなどの名前が保存できるようになります。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
1点のみ追加でコメントいたしましたので、ご確認いただけますと幸いです。

修正後、再度レビュー依頼フォームのご提出ください。
お手数をお掛けしますが、よろしくお願いいたします。

ken-forest committed:
userモデルのバリデーションを変更

具体的には、last_nameとfirst_nameのバリデーションを
「/\A[ぁ-んァ-ヶ一-龥々ー]+\z/ 」
に変更。

ken-forest commented:
ご指摘いただきありがとうございました。
非常に勉強になりました！
お手数ですが、引き続きレビューをよろしくお願いします。

zonodance reviewed (commented):
レビュー依頼ありがとうございます！

`LGTM`とさせていただきます！ユーザ管理機能の実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！

ken-forest commented:
レビューいただいた方々、
時間外での対応ありがとうございましたm(_ _)m
